src/key.js

```diff
@@ -771,6 +771,84 @@ Key.prototype.verifyAllUsers = async function(keys) {
   return results;
 };
 
+/**
+ * generate a new subkey to a Key
+ * @param  {Object} options   (optional) options for creating subkey, e.g. {sign: true, passphrase: '123', algorithm:..., curve:...}
+ *
+ * @returns {Promise<module:key.SubKey>} return the subkey if successful.
+ * @async
+ */
+Key.prototype.generateSubkey = async function(options){
+  const defaultOptions = {};
+  if (this.primaryKey.algorithm.indexOf('rsa') === 0) {
+    defaultOptions.numBits = 2048;
+  }
+  else {
+    defaultOptions.curve = enums.curve.ed25519;
+  }
+
+  options = sanitizeKeyOptions(options, defaultOptions);
+  const secretSubKeyPacket = await generateSecretSubkey(options);
+  const result = await this.addSubkey(secretSubKeyPacket, options);
+  return result;
+};
+
+/**
+ * add a subkey packet to a Key
+ * @param {module:packet.PublicSubkey|module:packet.SecretSubkey} the subkey packet to add
+ * @param  {Object} options   (optional) options for subkey, e.g. {sign: true, passphrase: '123'}
+ *
+ * @returns {Promise<module:key.SubKey>} return the subkey if successful.
+ * @async
+ */
+Key.prototype.addSubkey = async function(subkeyPacket, options){
+  if (!this.isPrivate()) {
+    throw new Error("Nothing to addSubkey in a public key");
+  }
+  if (!options) options = {};
+  let needRelocked;
+
+  const secretKeyPacket = this.primaryKey;
+
+  if (!secretKeyPacket.isDecrypted) {
+    try {
+      await secretKeyPacket.decrypt(options.passphrase);
+      needRelocked = true;
+    } catch (err) {
+      throw new Error('Key not decrypted');
+    }
+  }
+
+  if (options.passphrase) {
+    await subkeyPacket.encrypt(options.passphrase);
+  }
+
+  //sign the subkey here
+  const subkeySignatureOpts = {};
+  subkeySignatureOpts.signatureType = enums.signature.subkey_binding;
+  subkeySignatureOpts.keyFlags = options.sign ? enums.keyFlags.sign_data : [enums.keyFlags.encrypt_communication | enums.keyFlags.encrypt_storage];
+  if (options.keyExpirationTime > 0) {
+    subkeySignatureOpts.keyExpirationTime = options.keyExpirationTime;
+    subkeySignatureOpts.keyNeverExpires = false;
+  }
+  const dataToSign = {key: secretKeyPacket, bind: subkeyPacket};
+  const subkeySignaturePacket = await createSignaturePacket(
+    dataToSign, this, secretKeyPacket,
+    subkeySignatureOpts, options.date, options.userId
+  );
+  if (options.passphrase) {
+    subkeyPacket.clearPrivateParams();
+  }
+  const subKey = new SubKey(subkeyPacket);
+  subKey.bindingSignatures.push(subkeySignaturePacket);
+  this.subKeys.push(subKey);
+  if (needRelocked) {
```

Comment:
This is assuming that the secret key and subkey use the same encryption password. While this is true in most cases, it is not true as a general rule (see the way key generation handles/allows for multiple passphrases for each subkey). 
(This is also relevant to your line 815)

Reply:
No.  It just is reuse the `options.passphrase`.  The `needRelocked` only used to relock security key.

* Case 1:  Set the same password to subkey which is your case.
* Case 2：Set the different password to the subkey, not the same with secret key.
  * Case 2.1: It should raise error if the secret key is encrypted.
    * because the password is wrong for the secret key. So the user must decrypt it first.
  * Case 2.2: It will only encrypt the subkey with the password if the secret key is decrypted.
    * because the needRelocked is false: undefined.